javascript_nodejs/10.prompt-validations/dialogs/onboard/index.js

```diff
@@ -0,0 +1,114 @@
+const { ComponentDialog, WaterfallDialog } = require('botbuilder-dialogs');
+const moment = require('moment');
+
+// Import a few specialized prompt classes.
+const NamePrompt = require('../../prompts/namePrompt');
+const AgePrompt = require('../../prompts/agePrompt');
+const DOBPrompt = require('../../prompts/dobPrompt');
+const ColorPrompt = require('../../prompts/colorPrompt');
+
+const START_DIALOG = 'start';
+const HELLO_USER = 'welcome_back';
+
+const GET_NAME_PROMPT = 'namePrompt';
+const GET_AGE_PROMPT = 'agePrompt';
+const GET_DOB_PROMPT = 'dobPrompt';
+const GET_COLOR_PROMPT = 'colorPrompt';
+
+const USER_NAME_PROPERTY = 'name';
+const AGE_PROPERTY = 'age';
+const DOB_PROPERTY = 'dob';
+const COLOR_PROPERTY = 'color';
+
+class OnboardingDialog extends ComponentDialog {
+
+    /**
+     * 
+     * @param {string} dialogId A unique identifier for this dialog.
+     * @param {BotStatePropertyAccessor} userProfile A property used to store the user's name.
+     */
+    constructor (dialogId, userProfile) {
+        super(dialogId);
+
+        // Create a dialog flow that captures a series of values from a user
+        this.addDialog(new WaterfallDialog(START_DIALOG,[
+            // If a user name is already set, switch to the HELLO_USER dialog.
+            // Otherwise, continue with collecting values from the user.
+            async (dc, step) => {
+                const user = await userProfile.get(dc.context, {});
+                if (user.name) {
+                    return dc.replace(HELLO_USER);
```

Comment:
If a user answers the name prompt and then cancels out on subsequent prompts, it doesn't look like they'll ever be able set the rest of their user info.

Reply:
Since the user object is only updated at the end of the flow, there is no way to have partial info, thus checking for the name field should be enough.  Thoughts?

Reply:
Ahh I see, that's correct! Sorry I didn't see that 😅 

On that note then, can you add a note to the second waterfall step, where it stores the value in the `WaterfallStepContext` so we can tell users that there is a data store that lasts for the context/duration of the WaterfallDialog?

Reply:
Yes!